Devdocs intro (#5765)
gammapy/gammapy: docs/development/intro.rst

Get set up
==========

The first steps are basically identical to
https://ctapipe.readthedocs.io/en/latest/developer-guide/getting-started.html (until
section *Setting up the development environment*) and
http://astropy.readthedocs.io/en/latest/development/workflow/get_devel_version.html
(up to *Create your own private workspace*). The following is a quick summary of
commands to set up an environment for Gammapy development:

> First of all, fork the gammapy repository, https://github.com/gammapy/gammapy, to your personal GitHub. If you have not already done so, copy your public SSH key (normally under ``$HOME/.ssh/*.pub``) to your profile: **Settings/SSH and GPG keys/New SSH key**. Select **Authentication Key** as the key type. If you do not have a SSH key, you may create it following https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent.

Reviewer: ```suggestion
Firstly, fork the `gammapy repository <https://github.com/gammapy/gammapy>`__ to your personal GitHub. If you have not already done so, copy your public SSH key (normally under ``$HOME/.ssh/*.pub``) to your GitHub profile: **Settings/SSH and GPG keys/New SSH key**. Select **Authentication Key** as the key type. If you do not have an SSH key, you may create it using the following: `generate ssh key <https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent>`__.
```

Get set up
==========

.. note::
        :class: sphx-glr-download-link-note

        This section is based on the
        `instructions <https://ctapipe.readthedocs.io/en/latest/developer-guide/getting-started.html>`_

>         for setting up ``ctapipe``.

Reviewer 2: Maybe, one can put the documentation link: https://ctapipe.readthedocs.io/en/latest/

your personal GitHub. If you have not already done so, copy your public SSH key
(normally under ``$HOME/.ssh/*.pub``) to your GitHub profile by going into your
GitHub settings and doing **SSH and GPG keys/New SSH key** and selecting **Authentication Key**
as the key type. If you do not have a SSH key, you may create it following the instructions in
`this page <https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent>`_.

Go to the location in your machine where you want to code and clone the gammapy
repository from your GitHub:

.. code-block:: bash

    cd [your-working-path]
    git clone git@github.com:[your-github-username]/gammapy.git

Create the conda environment:

.. code-block:: bash

    # Fork the gammapy repository on GitHub, https://github.com/gammapy/gammapy
    cd code # Go somewhere on your machine where you want to code
    git clone https://github.com/[your-github-username]/gammapy.git
    cd gammapy
    conda env create -f environment-dev.yml

    # To speed up the environment solving you can use mamba instead of conda

>     # mamba env create -f environment-dev.yml

Reviewer 2: Why removing mamba?

    conda activate gammapy-dev

    # for conda versions <4.4.0 you may have to execute
    # 'source activate gammapy-dev' instead
    git remote add gammapy git@github.com:gammapy/gammapy.git

>     git remote rename origin [your-user-name]

Reviewer 2: Maybe one can precise that one can do that, in comment... It is quite used by some

If you are already working with the development version environment and you want
to update it with the contents of `environment-dev.yml` run the following command:

> `Mamba <https://mamba.readthedocs.io/>`__ is an alternative package manager that offers higher installation

Reviewer 2: Mamba is completely removed? Why? To be kept, I think

speed and more reliable environment solutions.
.. code-block:: bash

It is also common to stick with the name ``origin`` for your repository and to
use ``upstream`` for the repository you forked from. In any case, you can use

> ``$ git remote -v`` to list all your configured remotes.

Reviewer 2: These comments can be useful, outside the bash session

>     git branch [branch-name]

Reviewer 3: this is shown below. So fine like this.

Get set up
==========

This section is based on the `instructions <https://ctapipe.readthedocs.io/en/latest/>`_

> for setting up ``ctapipe``.

Reviewer 2: Defined what is ctapipe... This is unknown for many people

Firstly, fork the `gammapy repository <https://github.com/gammapy/gammapy>`_ to
your personal GitHub. If you have not already done so, copy your public SSH key
(normally under ``$HOME/.ssh/*.pub``) to your GitHub profile by going into your
GitHub settings and doing **SSH and GPG keys/New SSH key** and selecting **Authentication Key**
as the key type. If you do not have a SSH key, you may create it following the instructions in
`this page <https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent>`_.

> Go to the location in your machine where you want to code and clone the gammapy

Reviewer: ```suggestion
Go to the location on your machine where you want to clone the gammapy
```

.. code-block:: bash

    conda env update --file environment-dev.yml --prune

Add the gammapy repository as your upstream remote:

.. code-block:: bash

    git remote add upstream git@github.com:gammapy/gammapy.git

You can use ``git remote -v`` to list all your configured remotes.
If you followed the steps above, you should see something like this:

.. code-block:: bash

    origin	git@github.com:[your-github-username]/gammapy.git (fetch)
    origin	git@github.com:[your-github-username]/gammapy.git (push)
    upstream	git@github.com:gammapy/gammapy.git (fetch)
    upstream	git@github.com:gammapy/gammapy.git (push)

Make your ``main`` branch track the gammapy repository:

.. code-block:: bash

    git fetch upstream
    git branch --set-upstream-to=upstream/main

When developing Gammapy you never want to work on the ``main`` branch, but always
on a dedicated feature branch. You can create a new branch with:

.. code-block:: bash

    git branch [branch-name]

To switch between different branches:

.. code-block:: bash

    git checkout [branch-name]

You can also do both things at once as:

.. code-block:: bash

    git checkout -b [branch-name]

> To *activate* your development version (branch) of Gammapy in your environment run:

Reviewer: ```suggestion
To *activate* your development version (branch) of Gammapy in your environment:
```

.. code-block:: bash

    python -m pip install -e .

This is necessary to compile the few Cython code  (``*.pyx``). If you skip
this step, some imports depending on Cython code will fail. If you want to remove the generated
files run ``make clean``.

Every time that you want to ``push`` your local branch to your GitHub repository you should run:

.. code-block:: bash

    git push origin [branch-name]

Download the Gammapy datasets and declare the corresponding bash environment variable:

.. code-block:: bash

    gammapy download datasets --out GAMMAPY_DATA
    export GAMMAPY_DATA=$PWD/GAMMAPY_DATA/dev

You may include the line ``export GAMMAPY_DATA=$PWD/GAMMAPY_DATA/dev`` to your ``$HOME/.bashrc``

> or ``$HOME/.bash_profile``, so that the variable is available in every terminal session.

Author: Good point! Probably better to define the variable only within the environment, and not globally.

> or ``$HOME/.bash_profile``, so that the variable is available in every terminal session.

Reviewer 2: It is up to people, I would say... The most important is anyway to mention to set up this global env variable;)
